src/lib/datatypes/sol-str-slice.c

```diff
@@ -48,16 +48,18 @@ sol_str_slice_to_int(const struct sol_str_slice s, int *value)
     return 0;
 }
 
-SOL_API bool
+SOL_API char *
 sol_str_slice_contains(const struct sol_str_slice haystack, const struct sol_str_slice needle)
 {
-    return memmem(haystack.data, haystack.len, needle.data, needle.len) != NULL;
+    return memmem(haystack.data, haystack.len, needle.data, needle.len);
```

Comment:
you can do a very nice simple optimization here:

if (needle.len == 1) return memchr() else return memmem();